### PR TITLE
Add QM vocab inverse labels

### LINKED ISSUE #261: Add some inverse labels to QM vocabulary
For reporting across RM and QM domains, oslc:inverseLabel values on properties are useful. In particular, the following are suggested:
```

oslc_qm:affectedByChangeRequest         oslc:inverseLabel  "Affects Test Case Result" .
oslc_qm:blockedByChangeRequest          oslc:inverseLabel  "Blocks Test Case Execution Record" .
oslc_qm:executesTestScript              oslc:inverseLabel  "Executed By" .
oslc_qm:relatedChangeRequest            oslc:inverseLabel  "Related Test Artifact" .
oslc_qm:reportsOnTestCase               oslc:inverseLabel  "Reported By" .
oslc_qm:reportsOnTestPlan               oslc:inverseLabel  "Reported By" .
oslc_qm:runsTestCase                    oslc:inverseLabel  "Run By" .
oslc_qm:testsChangeRequest              oslc:inverseLabel  "Tested By Test Case" .
oslc_qm:usesTestCase                    oslc:inverseLabel  "Used By" .
oslc_qm:usesTestScript                  oslc:inverseLabel  "Used By" .
oslc_qm:validatesRequirement            oslc:inverseLabel  "Validated By" .
oslc_qm:validatesRequirementCollection  oslc:inverseLabel  "Validated By" .

```

### PR DESCRIPTION
**Pull request by berezovskyi**
Closes #261

### CODE REVIEW

**Comment by berezovskyi** [comment]
@ndjc I made this PR just to get the discussion going. I am not sure this is a good way to go with labels unless we also add `rdfs:domain` definitions. If we are unable to commit to a given RDFS domain for any uses of this property, I think using such inverse labels is a bit to assertive.

For example, who said that  `oslc_qm:blockedByChangeRequest` property can only be used on `oslc_qm:TestExecutionRecord`? I can easily imagine other things being blocked by a CR too. If something other than a `oslc_qm:TestExecutionRecord` is allowed to be the domain of the `oslc_qm:blockedByChangeRequest` property, the inverse label is not always going to be correct.

**Comment by ndjc** [comment]
I think the bigger problem is the name of the forward label, not the new inverse label! But of course it is really difficult for us to change that at this point.

Consider this one:

```
<http://open-services.net/ns/qm#affectedByChangeRequest>
     a                  rdf:Property ;
     rdfs:comment       "Change request that affects the Test Result. It is likely that the target resource will be an oslc_cm:ChangeRequest but that is not necessarily the case." ;
     rdfs:isDefinedBy   <http://open-services.net/ns/qm#> ;
     rdfs:label         "affectedByChangeRequest" .
     rdfs:label         "affectedByChangeRequest" ;
     oslc:inverseLabel  "Affects Test Case Result" .
```

This is a definition of a vocabulary term for QM - so it is all under QM control. The author of this vocabulary is saying that this property indicates the object affects the subject - and the subject is at least logically a test result (from the rdfs:comment). It also explicitly states that the object might not be an `oslc_cm:ChangeRequest`, even though the predicate name and the forward label strongly imply that.

Given that the predicate is defined by QM, and that the definition of the term implies the subject is a test result, the inverse label actually seems quite reasonable - while the original forward label is questionable.

I don't like the asymmetry of the camel case for the forward label and the title case with spaces for the inverse label. We should probably go back to the team suggesting this to see why the different conventions were used.

As for `rdfs:domain` and `rdfs:range`, the major problem we have seen in the past is that they tend not to mean what people think they mean. They tend to think they are suggestions, and they tend to think of them as being a set of equally valid alternatives if multiple values are given - so the expectation is that adding `rdfs:range <a> , <b>` means the object could be either `<a>` or `<b>`, whereas in fact it infers that any target is both an `<a>` and a `<b>`, as well as whatever other type it defines for itself.

I see no reason to object to either `rdfs:domain` or `rdfs:range` if the author of the vocabulary fully understands their meaning, and has though through the implications of their use. I use rdfs:range in a few places in the configuration management spec - for example:

```
oslc_config:baselineOfStream
    a rdf:Property ;
    rdfs:isDefinedBy oslc_config: ;
    rdfs:label "baselineOfStream" ;
    rdfs:range oslc_config:Stream ;
    rdfs:comment "The stream from which a baseline was made." .
```
I am happy for a reader to infer that any resource that is the object of this property is implied to be an OSLC configuration management stream, even if it is not explicitly defined as one - and that inference is in line with the predicate name that implies the object is a stream.

**Comment by berezovskyi** [comment]
Thank you, Nick.

When I see the property like `affectedByChangeRequest`, I imagine anything it points to IS a `ChangeRequest`. Now, you are right that it doesn't have to (as I mentioned many times, URIs shall be opaque). You are also right that over time, many people (including academics) have misused RDFS domain/range predicates. TopBraid literally includes a tool that takes (presumably) "backwards" RDFS domain/range definitions and turns them into SHACL shapes with proper domain/range constraints. I think having "global" SHACL shapes, for example, would have been a better solution (e.g. PropertyShape instances with a sh:targetSubjectsOf/sh:targetObjectsOf target definition, see http://book.validatingrdf.com/bookHtml011.html#sec135 for more details). I think we should consider a similar OSLC Shape definition and if OSLC Shapes do not allow property definitions outside resource shapes, we should strongly consider adopting SHACL (I will remind you that Lyo has a near-complete converter from OSLC Shapes to SHACL). 

Regarding the camel case vs non-camel case - I am not sure updating the rdfs:label would be a good idea, many tools show it as a property name. I think non-camel casing the inverse is also strange - camelCase implies there is an exact machine-readable copy of this definition, but there is not...

By the way, I read https://oslc-op.github.io/oslc-specs/specs/core/core-vocab.html#h3_invLabels and I think we should go for those inverse labels but in a lighter way without asserting the type of the object:

oslc_qm:blockedByChangeRequest
	rdfs:label "blockedByChangeRequest" ;
	oslc:inverseLabel "blocks" .

So, I suggest to keep it light in terms of types and let the subjects/object and their hover UIs tell the types. Let's limit properties to describing properties. So, it could look like this:

[PO-123](#) blocks [REQ-234](#)

This way, if `PO-123` is the ID for the purchasing order (trying to make up an unrelated domain), the user can hover over the link for a tooltip. The inverse label is still valid.

P.S. I am not entirely sure why do you think the forward label is set in stone - I think we can edit it as long as it aids understanding. I do not consider labels to have any semantic meaning and as such changes to them can hardly introduce backwards incompatibility.

**Comment by ndjc** [comment]
Aren't you arguing both sides of the case about changing labels here:

> Regarding the camel case vs non-camel case - I am not sure updating the rdfs:label would be a good idea, many tools show it as a property name. I think non-camel casing the inverse is also strange - camelCase implies there is an exact machine-readable copy of this definition, but there is not...

vs.

> P.S. I am not entirely sure why do you think the forward label is set in stone - I think we can edit it as long as it aids understanding. I do not consider labels to have any semantic meaning and as such changes to them can hardly introduce backwards incompatibility.

I'm also not entirely sure why I first reacted strongly against a change to the previously published label. I am certainly willing to discuss changes to it - and we'd need to discuss any potential change with tool vendors including IBM to see what impact it would have, if any.

File: specs/qm/quality-management-vocab.ttl [hunk 1]
```diff
@@ -69,67 +69,78 @@
...
+        rdfs:label        "reportsOnTestPlan" ;
+        oslc:inverseLabel  "Reported By" .
 
 <http://open-services.net/ns/qm#relatedChangeRequest>
         a                 rdf:Property ;
         rdfs:comment      "A related change request. It is likely that the target resource will be an oslc_cm:ChangeRequest but that is not necessarily the case." ;
         rdfs:isDefinedBy  <http://open-services.net/ns/qm#> ;
-        rdfs:label        "relatedChangeRequest" .
+        rdfs:label        "relatedChangeRequest" ;
+        oslc:inverseLabel  "Related Test Artifact" .
```
Comment (berezovskyi) [inline specs/qm/quality-management-vocab.ttl]: relates to

File: specs/qm/quality-management-vocab.ttl [hunk 2]
```diff
@@ -69,67 +69,78 @@
         a                 rdf:Property ;
         rdfs:comment      "Change request that affects the Test Result. It is likely that the target resource will be an oslc_cm:ChangeRequest but that is not necessarily the case." ;
         rdfs:isDefinedBy  <http://open-services.net/ns/qm#> ;
-        rdfs:label        "affectedByChangeRequest" .
+        rdfs:label        "affectedByChangeRequest" ;
+        oslc:inverseLabel  "Affects Test Case Result" .
```
Comment (berezovskyi) [inline specs/qm/quality-management-vocab.ttl]: affects

File: specs/qm/quality-management-vocab.ttl [hunk 3]
```diff
@@ -69,67 +69,78 @@
...
+        rdfs:label        "relatedChangeRequest" ;
+        oslc:inverseLabel  "Related Test Artifact" .
 
 <http://open-services.net/ns/qm#testsChangeRequest>
         a                 rdf:Property ;
         rdfs:comment      "Change Request tested by the Test Case. It is likely that the target resource will be an oslc_cm:ChangeRequest but that is not necessarily the case." ;
         rdfs:isDefinedBy  <http://open-services.net/ns/qm#> ;
-        rdfs:label        "testsChangeRequest" .
+        rdfs:label        "testsChangeRequest" ;
+        oslc:inverseLabel  "Tested By Test Case" .
```
Comment (berezovskyi) [inline specs/qm/quality-management-vocab.ttl]: tested by

File: specs/qm/quality-management-vocab.ttl [hunk 4]
```diff
@@ -69,67 +69,78 @@
...
+        rdfs:label        "validatesRequirementCollection" ;
+        oslc:inverseLabel  "Validated By" .
 
 <http://open-services.net/ns/qm#blockedByChangeRequest>
         a                 rdf:Property ;
         rdfs:comment      "Change Request that prevents execution of the Test Execution Record. It is likely that the target resource will be an oslc_cm:ChangeRequest but that is not necessarily the case." ;
         rdfs:isDefinedBy  <http://open-services.net/ns/qm#> ;
-        rdfs:label        "blockedByChangeRequest" .
+        rdfs:label        "blockedByChangeRequest" ;
+        oslc:inverseLabel  "Blocks Test Case Execution Record" .
```
Comment (berezovskyi) [inline specs/qm/quality-management-vocab.ttl]: blocks

**Comment by ndjc** [comment]
I have no further comments on this PR - I'm OK with whatever you decide on those inverse labels.

**Pull request merged**